File: js/ai_algorithms.js
class CloudGraph {
  constructor(width, groundY, rows = 7, columns = 4, rowGap = 68, jumpRange = 185) {
    this.width = width;
    this.groundY = groundY;
    this.rows = rows;
    this.columns = columns;
    this.rowGap = rowGap;
    this.jumpRange = jumpRange;
    this.nodes = [];
    this.rowLookup = {};
    this.rowNodes = [];
    this.rowLayouts = [];
    this.edges = [];

    this._generateNodes();
    this._buildEdges();
  }

  _generateNodes() {
    const startY = this.groundY - 70;
    this.nodes.push([Math.floor(this.width / 2), startY]);
    this.rowLookup[0] = 0;
    this.rowNodes.push([0]);
    this.rowLayouts.push([Math.floor(this.width / 2)]);

    let nodeId = 1;
    let previousPositions = [Math.floor(this.width / 2)];

    for (let row = 1; row < this.rows; row += 1) {
      const y = startY - row * this.rowGap;
      const rowNodes = [];
      const placedX = this._makeScatteredPositions(previousPositions);
      previousPositions = [...placedX];
      this.rowLayouts.push(placedX);

      for (const x of placedX) {
        this.nodes.push([x, y]);
        this.rowLookup[nodeId] = row;
        rowNodes.push(nodeId);
        nodeId += 1;
      }

      this.rowNodes.push(rowNodes);
    }
  }

  _buildEdges() {
    this.edges = Array.from({ length: this.nodes.length }, () => []);

    for (let row = 0; row < this.rows - 1; row += 1) {
      for (const nodeId of this.rowNodes[row]) {
        const [x1] = this.nodes[nodeId];
        const nextRow = row + 1;
        if (nextRow >= this.rows) {
          continue;
        }

        for (const neighborId of this.rowNodes[nextRow]) {
          const [x2] = this.nodes[neighborId];
          if (Math.abs(x2 - x1) <= this.jumpRange) {
            this.edges[nodeId].push(neighborId);
          }
        }
      }
    }
  }

  addRowsAbove(count = 1) {
    for (let index = 0; index < count; index += 1) {
      const topY = Math.min(...this.nodes.map((node) => node[1]));
      const y = topY - this.rowGap;
      const newRowIndex = this.rowNodes.length;
      const previousPositions = this.rowLayouts.length ? this.rowLayouts[this.rowLayouts.length - 1] : [Math.floor(this.width / 2)];
      const placedX = this._makeScatteredPositions(previousPositions);
      this.rowLayouts.push(placedX);
      const rowNodes = [];

      for (const x of placedX) {
        const nodeId = this.nodes.length;
        this.nodes.push([x, y]);
        this.rowLookup[nodeId] = newRowIndex;
        rowNodes.push(nodeId);
      }

      this.rowNodes.push(rowNodes);
    }

    this._buildEdges();
  }

  _makeScatteredPositions(previousPositions) {
    const margin = 60;
    const allowedMin = margin;
    const allowedMax = this.width - margin;
    const minSpacing = 85;

    const reachable = (xValue) => previousPositions.some((previousX) => Math.abs(xValue - previousX) <= this.jumpRange);
    const clamp = (xValue) => Math.max(allowedMin, Math.min(allowedMax, Math.trunc(xValue)));
    const farEnough = (candidate, positions) => positions.every((existing) => Math.abs(candidate - existing) >= minSpacing);

    const positions = [];

    for (const parentX of previousPositions) {
      let candidate = null;
      for (let attempt = 0; attempt < 20; attempt += 1) {
        const delta = this._randomInt(-(this.jumpRange - 45), this.jumpRange - 45);
        const trial = clamp(parentX + delta);
        if (reachable(trial) && farEnough(trial, positions)) {
          candidate = trial;
          break;
        }
      }

      if (candidate === null) {
        const direction = Math.random() < 0.5 ? -1 : 1;
        candidate = clamp(parentX + direction * Math.max(60, Math.floor(this.jumpRange / 2) - 15));
      }

      if (farEnough(candidate, positions)) {
        positions.push(candidate);
      }
    }

    const targetCount = this._randomInt(Math.max(2, previousPositions.length), Math.max(2, this.columns));
    let attempts = 0;
    while (positions.length < targetCount && attempts < 200) {
      attempts += 1;
      let candidate;

      if (Math.random() < 0.5) {
        const parentX = previousPositions[this._randomInt(0, previousPositions.length - 1)];
        candidate = clamp(parentX + this._randomInt(-(this.jumpRange - 45), this.jumpRange - 45));
      } else {
        candidate = this._randomInt(allowedMin, allowedMax);
      }

      if (!reachable(candidate)) {
        const nearestPrev = previousPositions.reduce((best, current) => Math.abs(current - candidate) < Math.abs(best - candidate) ? current : best, previousPositions[0]);
        const direction = candidate > nearestPrev ? -1 : 1;
        candidate = clamp(nearestPrev + direction * this._randomInt(55, Math.max(60, this.jumpRange - 35)));
      }

      if (farEnough(candidate, positions)) {
        positions.push(candidate);
      }
    }

    const unique = [...new Set(positions.map((position) => Math.trunc(position)))].sort((left, right) => left - right);

    if (unique.length < 2) {
      const anchor = unique.length ? unique[0] : previousPositions[0];
      const fallback = clamp(anchor + (Math.random() < 0.5 ? -1 : 1) * Math.max(80, Math.floor(this.jumpRange / 2)));
      if (farEnough(fallback, unique)) {
        unique.push(fallback);
      }
    }

    return unique.slice(0, Math.max(2, this.columns)).sort((left, right) => left - right);
  }

  _randomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  getNodePosition(nodeId) {
    return this.nodes[nodeId];
  }

  getReachableNodes(nodeId) {
    return this.edges[nodeId] || [];
  }

  getRow(nodeId) {
    return this.rowLookup[nodeId];
  }

  get startNode() {
    return 0;
  }

  get goalRow() {
    return this.rows - 1;
  }
}

class BFSSolver {
  isSequenceSurvivable(cloudGraph, route) {
    if (!route || !route.length) {
      return false;
    }

    const goal = route[route.length - 1];
    const visited = new Set([route[0]]);
    const queue = [route[0]];

    while (queue.length) {
      const current = queue.shift();
      if (current === goal) {
        return true;
      }

      for (const neighbor of cloudGraph.getReachableNodes(current)) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push(neighbor);
        }
      }
    }

    return false;
  }
}

class IDSSequenceGenerator {
  constructor(cloudGraph) {
    this.cloudGraph = cloudGraph;
    this.bfsSolver = new BFSSolver();
    this.startCloud = cloudGraph.startNode;
  }

  generateSequence(targetDepth, maxAttempts = 30) {
    const maxDepth = Math.min(targetDepth, this.cloudGraph.goalRow);

    for (let depth = 1; depth <= maxDepth; depth += 1) {
      const route = this._depthLimitedSearch(depth);
      if (route && this.bfsSolver.isSequenceSurvivable(this.cloudGraph, route)) {
        return this._buildSequence(route);
      }
    }

    const route = this._fallbackRoute();
    return this._buildSequence(route);
  }

  _depthLimitedSearch(depthLimit) {
    const targetRow = Math.min(depthLimit, this.cloudGraph.goalRow);

    const recurse = (nodeId, depthRemaining, path, visited) => {
      if (depthRemaining === 0) {
        if (this.cloudGraph.getRow(nodeId) >= targetRow) {
          return path;
        }
        return null;
      }

      const neighbors = [...this.cloudGraph.getReachableNodes(nodeId)].sort((left, right) => {
        const leftRow = this.cloudGraph.getRow(left);
        const rightRow = this.cloudGraph.getRow(right);
        const center = Math.floor(this.cloudGraph.width / 2);
        const leftDistance = Math.abs(this.cloudGraph.getNodePosition(left)[0] - center);
        const rightDistance = Math.abs(this.cloudGraph.getNodePosition(right)[0] - center);
        return leftRow - rightRow || leftDistance - rightDistance;
      });

      for (const neighborId of neighbors) {
        if (visited.has(neighborId)) {
          continue;
        }

        visited.add(neighborId);
        const result = recurse(neighborId, depthRemaining - 1, [...path, neighborId], visited);
        if (result) {
          return result;
        }
        visited.delete(neighborId);
      }

      return null;
    };

    return recurse(this.startCloud, depthLimit, [this.startCloud], new Set([this.startCloud]));
  }

  _buildSequence(route) {
    return {
      route,
      hazards: this._buildHazards(route),
      route_depth: route.length - 1,
    };
  }

  _buildHazards(route) {
    const routeSet = new Set(route);
    const routeTopRow = this.cloudGraph.getRow(route[route.length - 1]);
    const candidates = [];

    for (let nodeId = 0; nodeId < this.cloudGraph.nodes.length; nodeId += 1) {
      if (!routeSet.has(nodeId) && this.cloudGraph.getRow(nodeId) <= routeTopRow) {
        candidates.push(nodeId);
      }
    }

    this._shuffle(candidates);
    const hazards = [];

    const limit = Math.min(candidates.length, Math.max(2, route.length) + 1);
    for (let index = 0; index < limit; index += 1) {
      const nodeId = candidates[index];
      const row = this.cloudGraph.getRow(nodeId);
      const speed = 3.0 + row * 0.45;
      const base = 0.5;
      const spacing = 0.8;
      const jitter = this._randomRange(-0.2, 0.3);
      const spawnTime = Math.max(0.2, Number((base + (index + 1) * spacing + jitter).toFixed(2)));
      hazards.push([nodeId, spawnTime, speed]);
    }

    hazards.sort((left, right) => left[1] - right[1]);
    return hazards;
  }

  _fallbackRoute() {
    const route = [this.startCloud];
    const visited = new Set([this.startCloud]);
    let current = this.startCloud;

    while (this.cloudGraph.getRow(current) < this.cloudGraph.goalRow) {
      const neighbors = this.cloudGraph.getReachableNodes(current);
      let nextNode = null;

      for (const neighbor of neighbors) {
        if (!visited.has(neighbor)) {
          nextNode = neighbor;
          break;
        }
      }

      if (nextNode === null) {
        break;
      }

      route.push(nextNode);
      visited.add(nextNode);
      current = nextNode;
    }

    return route;
  }

  _shuffle(items) {
    for (let index = items.length - 1; index > 0; index -= 1) {
      const swapIndex = this._randomInt(0, index);
      [items[index], items[swapIndex]] = [items[swapIndex], items[index]];
    }
  }

  _randomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }

  _randomRange(min, max) {
    return Math.random() * (max - min) + min;
  }
}

File: ai_algorithms.py
"""
AI search layer for Cloud Catchers.

This module contains the three classical algorithms requested for the game:
- IDS: builds a short playable ascent route by increasing the search depth.
- DLS: performs the bounded traversal used by IDS.
- BFS: validates that the generated route is actually reachable.
"""

import random
from collections import deque


class CloudGraph:
    """
    Graph of cloud platforms arranged as an upward path network.

    Nodes are stored row-by-row from the starting cloud upward so the frog can
    climb vertically through the level.
    """

    def __init__(self, width, ground_y, rows=7, columns=4, row_gap=68, jump_range=185):
        self.width = width
        self.ground_y = ground_y
        self.rows = rows
        self.columns = columns
        self.row_gap = row_gap
        self.jump_range = jump_range
        self.nodes = []
        self.row_lookup = {}
        self.row_nodes = []
        self.row_layouts = []
        self.edges = []

        self._generate_nodes()
        self._build_edges()

    def _generate_nodes(self):
        start_y = self.ground_y - 70

        # Row 0 is the starting cloud the frog begins on.
        self.nodes.append((self.width // 2, start_y))
        self.row_lookup[0] = 0
        self.row_nodes.append([0])
        self.row_layouts.append([self.width // 2])

        node_id = 1
        previous_positions = [self.width // 2]

        # Create each row as a scattered set of clouds, similar to Doodle Jump.
        for row in range(1, self.rows):
            y = start_y - row * self.row_gap
            row_nodes = []

            placed_x = self._make_scattered_positions(previous_positions)
            previous_positions = placed_x[:]
            self.row_layouts.append(placed_x)

            for x in placed_x:
                self.nodes.append((x, y))
                self.row_lookup[node_id] = row
                row_nodes.append(node_id)
                node_id += 1

            self.row_nodes.append(row_nodes)

    def _build_edges(self):
        self.edges = [[] for _ in range(len(self.nodes))]

        for row in range(self.rows - 1):
            for node_id in self.row_nodes[row]:
                x1, _ = self.nodes[node_id]

                # Allow jumps only to the immediate next row above.
                next_row = row + 1
                if next_row < self.rows:
                    for neighbor_id in self.row_nodes[next_row]:
                        x2, _ = self.nodes[neighbor_id]
                        if abs(x2 - x1) <= self.jump_range:
                            self.edges[node_id].append(neighbor_id)

    def add_rows_above(self, count=1):
        """Grow the cloud graph upward by adding new random rows above the current top."""
        for _ in range(count):
            top_y = min(y for _, y in self.nodes)
            y = top_y - self.row_gap
            new_row_index = len(self.row_nodes)
            previous_positions = self.row_layouts[-1] if self.row_layouts else [self.width // 2]
            placed_x = self._make_scattered_positions(previous_positions)
            self.row_layouts.append(placed_x)
            row_nodes = []

            for x in placed_x:
                node_id = len(self.nodes)
                self.nodes.append((x, y))
                self.row_lookup[node_id] = new_row_index
                row_nodes.append(node_id)

            self.row_nodes.append(row_nodes)

        # Rebuild adjacency so new rows can be used immediately.
        self._build_edges()

    def _make_scattered_positions(self, previous_positions):
        """Create scattered cloud positions for the next row while preserving reachability."""
        margin = 60
        allowed_min = margin
        allowed_max = self.width - margin
        min_spacing = 85

        def reachable(x_value):
            return any(abs(x_value - previous_x) <= self.jump_range for previous_x in previous_positions)

        def clamp(x_value):
            return max(allowed_min, min(allowed_max, int(x_value)))

        def far_enough(candidate, positions):
            return all(abs(candidate - existing) >= min_spacing for existing in positions)

        positions = []

        # Guarantee one reachable child for every cloud in the row below.
        for parent_x in previous_positions:
            candidate = None
            for _ in range(20):
                delta = random.randint(-(self.jump_range - 45), self.jump_range - 45)
                trial = clamp(parent_x + delta)
                if reachable(trial) and far_enough(trial, positions):
                    candidate = trial
                    break

            if candidate is None:
                direction = -1 if random.random() < 0.5 else 1
                candidate = clamp(parent_x + direction * max(60, self.jump_range // 2 - 15))

            if far_enough(candidate, positions):
                positions.append(candidate)

        # Add a few extra scattered clouds, but keep them reachable from the row below.
        target_count = random.randint(max(2, len(previous_positions)), max(2, self.columns))
        attempts = 0
        while len(positions) < target_count and attempts < 200:
            attempts += 1
            if random.random() < 0.5:
                parent_x = random.choice(previous_positions)
                candidate = clamp(parent_x + random.randint(-(self.jump_range - 45), self.jump_range - 45))
            else:
                candidate = random.randint(allowed_min, allowed_max)

            if not reachable(candidate):
                nearest_prev = min(previous_positions, key=lambda previous_x: abs(previous_x - candidate))
                direction = -1 if candidate > nearest_prev else 1
                candidate = clamp(nearest_prev + direction * random.randint(55, max(60, self.jump_range - 35)))

            if far_enough(candidate, positions):
                positions.append(candidate)

        positions = sorted(dict.fromkeys(int(position) for position in positions))

        # Keep at least two clouds so the climb feels continuous.
        if len(positions) < 2:
            anchor = positions[0] if positions else previous_positions[0]
            fallback = clamp(anchor + random.choice([-1, 1]) * max(80, self.jump_range // 2))
            if far_enough(fallback, positions):
                positions.append(fallback)

        return sorted(positions[:max(2, self.columns)])

    def get_node_position(self, node_id):
        return self.nodes[node_id]

    def get_reachable_nodes(self, node_id):
        return self.edges[node_id]

    def get_row(self, node_id):
        return self.row_lookup[node_id]

    @property
    def start_node(self):
        return 0

    @property
    def goal_row(self):
        return self.rows - 1


class BFSSolver:
    """
    BFS validation layer.

    The generated route is accepted only when BFS can confirm the frog can
    still reach the target cloud chain in the cloud graph.
    """

    def is_sequence_survivable(self, cloud_graph, route):
        if not route:
            return False

        goal = route[-1]
        visited = {route[0]}
        queue = deque([route[0]])

        while queue:
            current = queue.popleft()
            if current == goal:
                return True

            for neighbor in cloud_graph.get_reachable_nodes(current):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

        return False


class IDSSequenceGenerator:
    """
    IDS controller for building short playable ascents.

    IDS increases the allowed depth one step at a time, DLS performs the
    bounded search at each depth, and BFS validates the final candidate.
    """

    def __init__(self, cloud_graph):
        self.cloud_graph = cloud_graph
        self.bfs_solver = BFSSolver()
        self.start_cloud = cloud_graph.start_node

    def generate_sequence(self, target_depth, max_attempts=30):
        max_depth = min(target_depth, self.cloud_graph.goal_row)

        # IDS implementation: keep increasing the depth limit until a route is found.
        for depth in range(1, max_depth + 1):
            route = self._depth_limited_search(depth)
            if route and self.bfs_solver.is_sequence_survivable(self.cloud_graph, route):
                return self._build_sequence(route)

        # Fallback sequence used only when the search space is unusually tight.
        route = self._fallback_route()
        return self._build_sequence(route)

    def _depth_limited_search(self, depth_limit):
        target_row = min(depth_limit, self.cloud_graph.goal_row)

        # DLS implementation: explore only up to the current depth limit.
        def recurse(node_id, depth_remaining, path, visited):
            if depth_remaining == 0:
                if self.cloud_graph.get_row(node_id) >= target_row:
                    return path
                return None

            neighbors = sorted(
                self.cloud_graph.get_reachable_nodes(node_id),
                key=lambda neighbor_id: (
                    self.cloud_graph.get_row(neighbor_id),
                    abs(self.cloud_graph.get_node_position(neighbor_id)[0] - self.cloud_graph.width // 2),
                ),
            )

            for neighbor_id in neighbors:
                if neighbor_id in visited:
                    continue
                visited.add(neighbor_id)
                result = recurse(neighbor_id, depth_remaining - 1, path + [neighbor_id], visited)
                if result:
                    return result
                visited.remove(neighbor_id)

            return None

        return recurse(self.start_cloud, depth_limit, [self.start_cloud], {self.start_cloud})

    def _build_sequence(self, route):
        hazards = self._build_hazards(route)
        return {
            "route": route,
            "hazards": hazards,
            "route_depth": len(route) - 1,
        }

    def _build_hazards(self, route):
        route_set = set(route)
        route_top_row = self.cloud_graph.get_row(route[-1])
        candidates = [
            node_id
            for node_id in range(len(self.cloud_graph.nodes))
            if node_id not in route_set and self.cloud_graph.get_row(node_id) <= route_top_row
        ]

        random.shuffle(candidates)
        hazards = []

        # Create more overlapping hazards early so at least two hazards appear
        for index, node_id in enumerate(candidates[: max(2, len(route)) + 1], start=1):
            row = self.cloud_graph.get_row(node_id)
            speed = 3.0 + row * 0.45
            # Spawn times biased toward early overlap; small randomness added
            base = 0.5
            spacing = 0.8
            jitter = random.uniform(-0.2, 0.3)
            spawn_time = max(0.2, round(base + index * spacing + jitter, 2))
            hazards.append((node_id, spawn_time, speed))

        hazards.sort(key=lambda item: item[1])
        return hazards

    def _fallback_route(self):
        route = [self.start_cloud]
        current = self.start_cloud
        visited = {self.start_cloud}

        while self.cloud_graph.get_row(current) < self.cloud_graph.goal_row:
            neighbors = self.cloud_graph.get_reachable_nodes(current)
            next_node = None
            for neighbor in neighbors:
                if neighbor not in visited:
                    next_node = neighbor
                    break

            if next_node is None:
                break

            route.append(next_node)
            visited.add(next_node)
            current = next_node

        return route